Update Core Data Model by adding `customFields` to-many relationship from `Product` to `MetaData`

pmusolino opened the pull request:
# DO NOT MERGE UNTIL https://github.com/woocommerce/woocommerce-ios/pull/13814 IS MERGED FIRST

Closes: #13832

## Description
This PR introduces some updates to the existing Core Data model 116 by adding `customFields` to-many relationship from `Product` to `MetaData`. It also includes tests to ensure the correctness of these changes and updates documentation regarding the migrations.

The changes are still not in use, since the Yosemite layer is missing and will be implemented in a future PR.

## Summary of changes
- **MetaData+CoreDataProperties.swift**:
  - Adds a new `product` inverse relationship to the `MetaData` entity.

- **Product+CoreDataProperties.swift**:
  - Adds a new `customFields` relationship to the `Product` entity.
  - Implements generated accessors for managing `customFields`.

- **StorageType+Extensions.swift**:
  - Renames the method `loadMetaData` to `loadOrderMetaData`.
  - Adds a new method `loadProductMetaData` to retrieve `MetaData` associated with a `Product`.

- **StorageTypeExtensionsTests.swift**:
  - Updates tests to reflect the method name change from `loadMetaData` to `loadOrderMetaData`.
  - Adds a new test `test_loadProductMetaData_by_siteID_metadataID` to verify loading `MetaData` associated with a `Product`.

- **OrdersUpsertUseCase.swift**:
  - Updates the logic to use `loadOrderMetaData` when handling order custom fields (both Order and Product entities now are in relationship with MetaData).

- **OrdersUpsertUseCaseTests.swift**:
  - Updates tests to reflect the use of `loadOrderMetaData`.

- **MigrationTests.swift**:
  - Adds a new test `test_migrating_from_115_to_116_adds_customFields_property_to_Product` to verify migration from model 115 to 116 correctly adds the `customFields` property to the `Product`.

- **MIGRATIONS.md**:
  - Updates documentation to reflect the new relationship from `Product` to `MetaData`.

- **Model 116.xcdatamodeld/contents**:
  - Updates the Core Data model to include the new relationships and properties for `Product` and `MetaData`.

## Testing
- Run all existing unit tests to ensure no regressions are introduced.
- Verify the new migration test to ensure the `customFields` property is correctly added during migration from model 115 to 116.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

pmusolino committed:
Update Storage model documentation and relationships
- Add `customFields` attribute to `Product` entity in `MIGRATIONS.md`
- Add `product` relationship to `customFields` in `WooCommerce.xcdatamodeld/Model 116.xcdatamodel/contents`
- Add `customFields` relationship to `MetaData` in `WooCommerce.xcdatamodeld/Model 116.xcdatamodel/contents`

pmusolino committed:
update: MIGRATIONS.md to reflect new relationship from `Product` to `OrderMetaData`.

pmusolino committed:
feat: add test for migrating from model 115 to 116
- Add `test_migrating_from_115_to_116_adds_customFields_property_to_Product` method in `MigrationTests.swift`
- Verify absence of `customFields` before migration
- Check product existence in model 115
- Confirm presence of `customFields` in migrated product
- Test adding custom fields to migrated product and validate relationships

pmusolino committed:
update: Core Data properties and methods for MetaData and Product

- Add `product` property to `MetaData` in `MetaData+CoreDataProperties.swift`
- Add `customFields` property and accessors to `Product` in `Product+CoreDataProperties.swift`
- Rename `loadMetaData` to `loadOrderMetaData` in `StorageType+Extensions.swift` and implement `loadProductMetaData`
- Update test cases to reflect new method names in `StorageTypeExtensionsTests.swift` and `OrdersUpsertUseCaseTests.swift`
- Adjust order handling in `OrdersUpsertUseCase.swift` to use `loadOrderMetaData`

selanthiraiyan reviewed (changes requested):
Instead of making changes to `Model 116`, I think we should follow the recommendation from p91TBi-3pE-p2 and add a new model version. 

> Once a model version is merged in `develop`, never modify it. Create a new model version instead. I pray that merge conflicts will happen if two people create the same new version number.

What do you think?

pmusolino commented:
@selanthiraiyan In the previous model (n°115) I see that two different contributions have been made, so I followed the same approach and forgot that discussion. But that suggestion makes sense, so I'll create a new model for this. 🙌

pmusolino committed:
Merge branch 'issue/13493-new-metadata-entity-in-product' into issue/13832-add-MetaData-to-Product-in-Storage

pmusolino committed:
revert: remove relationships from WooCommerce data model n°116
- Remove relationship "product" from entity
- Remove relationship "customFields" from entity

pmusolino committed:
feat: new core data model to version 117 and modified related files
- Update current version in WooCommerce.xcdatamodeld to Model 117
- Add new Model 117.xcdatamodel file with updated entity relationships
- Document new custom fields relationship in MIGRATIONS.md
- Add customFields relationship to Product in xcdatamodel contents
- Update test cases to include Model 117 in ManagedObjectModelsInventoryTests
- Modify migration test to reflect change from Model 116 to Model 117 in MigrationTests

pmusolino commented:
@selanthiraiyan changes made here f7438898e4119cdef26092f2aee054a7ee054016, ready for another check

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/222">20.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan commented on Storage/Storage/Tools/StorageType+Extensions.swift:
```diff
@@ -752,10 +752,18 @@ public extension StorageType {
         return firstObject(ofType: MetaData.self, matching: predicate)
     }
+
+    /// Retrieves the Stored Metadata for Order.
```
```suggestion
    /// Retrieves the Stored Metadata for a Product.
```

selanthiraiyan commented on Storage/Storage/Tools/StorageType+Extensions.swift:
```diff
@@ -752,10 +752,18 @@ public extension StorageType {
+
+    /// Retrieves the Stored Metadata for Order.
+    ///
+    func loadProductMetaData(siteID: Int64, metadataID: Int64) -> MetaData? {
+        let predicate = \MetaData.product?.siteID == siteID && \MetaData.metadataID == metadataID
+        return firstObject(ofType: MetaData.self, matching: predicate)
+    }
```
Will `metadataID` be unique across products and orders? 

I wonder what happens when a product and an order have metadata with the same ID. Will we end up loading order's meta data for a product and vice versa?

pmusolino replied:
I discussed this here C03L1NF1EA3/p1725454480621309.

On the web, MetaData for both orders and products are sharing the same table in the db `wp_postmeta`. However, both in iOS and in Android, we use the same approach: MetaData with the same IDs are allowed because it's possible that in the future meta data with the same ID will be allowed. For the moment, even if the table is the same, the endpoints for updating/creating metadata are linked just to orders or just to products, so it is impossible to create shared metadata. And in any case, the relationship from MetaData to Order/Product in CoreData is the following:
```
Order -> many MetaData
Product -> many MetaData
MetaData -> one Order
MetaData -> one Product
```

pmusolino replied:
I also made some changes here to be sure to fetch MetaData connected with a specific product or order: 4d0859bcf785c0bce5f55a2e3a32f94a26096a72

selanthiraiyan commented on Storage/Storage/Model/MetaData+CoreDataProperties.swift:
```diff
@@ -12,6 +12,7 @@ extension MetaData {
     @NSManaged public var value: String?
     @NSManaged public var metadataID: Int64
     @NSManaged public var order: Order?
+    @NSManaged public var product: Product?
```
I think we should update the comment in [this networking file](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Networking/Networking/Model/MetaData.swift#L4-L6).

pmusolino replied:
Good catch, done 508d94affe85bc22d1c96059ab34b5d6b9aa9b1c

pmusolino committed:
update: modify comment to specify that metadata can represent both Order and Product

pmusolino committed:
Update Storage/Storage/Tools/StorageType+Extensions.swift

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

selanthiraiyan reviewed (approved):
Thanks for the updates, Paolo! 🚀

pmusolino committed:
update: StorageType methods and tests to include `orderID` and `productID`

- Modify `loadOrderMetaData` and `loadProductMetaData` in `StorageType+Extensions.swift` to accept `orderID` and `productID` parameters.
- Update corresponding test methods in `StorageTypeExtensionsTests.swift` to reflect the new method signatures.
- Change `OrdersUpsertUseCase.swift` to use the updated `loadOrderMetaData` with `orderID`.
- Adjust `OrdersUpsertUseCaseTests.swift` to verify loading metadata with the new parameters.

pmusolino committed:
Merge commit '4b7fd10ec8903c8c68b7ed27716def74743ac925' into issue/13832-add-MetaData-to-Product-in-Storage

pmusolino committed:
refactor code for readability in OrdersUpsertUseCase.swift
- Break long line into multiple lines for better readability in the `handleOrderCustomFields` function.

pmusolino committed:
fix: trailing space
